### PR TITLE
Improve ADF Tool

### PR DESCRIPTION
**Pull request by skywalker-wu**
1. Support parallel processing for ADF Tool
2. Support cross region source and destination
3. Change to use streaming mode without overhead on disk

### CODE REVIEW

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirBlobConsumer.cs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+        public async Task CompleteAsync()
+        {
+            if (_currentStream.Length > 0)
+            {
+                await AddCurrentStreamToUploadTaskListAsync().ConfigureAwait(false);
+            }
+
+            await Task.WhenAll(_runningTasks).ConfigureAwait(false);
+            foreach (var task in _runningTasks.FindAll(t => t.IsCompleted))
```
Comment (deepak-ms) [inline src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirBlobConsumer.cs]: Shouldn't all be completed here (FindAll not needed)?

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirBlobConsumer.cs [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                }, TimeSpan.FromSeconds(BlockUploadTimeoutInSeconds), BlockUploadTimeoutRetryCount).ConfigureAwait(false);
+
+            ProgressHandler?.Report(stream.Length);
+            await stream.DisposeAsync().ConfigureAwait(false);
+        }
+
+        private static string GenerateBlockId(int index)
+        {
+            byte[] id = new byte[48]; // 48 raw bytes => 64 byte string once Base64 encoded
+            BitConverter.GetBytes(index).CopyTo(id, 0);
```
Comment (deepak-ms) [inline src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirBlobConsumer.cs]: not sure if i follow the reason for only 48 (<64) byte array

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirAzureConstants.cs [hunk 3]
```diff
@@ -0,0 +1,18 @@
+﻿namespace Fhir.Anonymizer.AzureDataFactoryPipeline.src
+{
+    public static class FhirAzureConstants
+    {
+        public const int KB = 1024;
+        public const int MB = KB * 1024;
+        public const int BlockBufferSize = 4 * MB;
+        public const int DefaultConcurrentCount = 3;
```
Comment (deepak-ms) [inline src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirAzureConstants.cs]: Should the concurrent counts be same as parallelism of anonymization?

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs [hunk 4]
```diff
@@ -120,43 +125,43 @@ private async Task AnonymizeBlobInJsonFormatAsync(BlobClient inputBlobClient, Bl
...
-                Console.WriteLine($"[{blobName}]: Succeeded in {processedCount} resources, failed in {processedErrorCount} resources in total.");
-            }
+                Interlocked.Add(ref processedCount, args.ProcessCompleted);
+                Interlocked.Add(ref processedErrorCount, args.ProcessFailed);
+                Interlocked.Add(ref consumedCount, args.ConsumeCompleted);
+
+                Console.WriteLine($"[{stopWatch.Elapsed.ToString()}][tid:{args.CurrentThreadId}]: {processedCount} Completed. {processedErrorCount} Failed. {consumedCount} consume completed.");
+            };
+
+            await executor.ExecuteAsync(CancellationToken.None, false, progress).ConfigureAwait(false);
```
Comment (deepak-ms) [inline src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs]: Can/Should we build resiliency against node crash in the middle of a large blob processing (say picking only unprocessed data on restart)?

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirBlobConsumer.cs (hunk 1 above)
Comment (skywalker-wu) [reply src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirBlobConsumer.cs]: Thanks, fixed.

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirBlobConsumer.cs (hunk 2 above)
Comment (skywalker-wu) [reply src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirBlobConsumer.cs]: It caused by the length limit for block is 64 bytes, and 48 bytes can be convert to a string with exactly 64 bytes.

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirAzureConstants.cs (hunk 3 above)
Comment (skywalker-wu) [reply src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirAzureConstants.cs]: Download or upload is in batch mode (4MB & 32MB), and also the download & upload is much faster(60MB/s for cross region) than processing(n * 100MB / min), looks like put 3 of them is enough? when executor is processing, there's at least 1~2 workers downloading and uploading.

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs (hunk 4 above)
Comment (skywalker-wu) [reply src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs]: Maybe we can add file level retry at ADF pipeline first? Use foreach activity to run anonymizer tool on single file and retry if it failed?

To achieve resume functionality, looks like we need somewhere for partial status, in batch mode, we can do this with additional storage table or ... If we change to use azure function, looks like we can easily leverage the partial status.

File: src/Fhir.Anonymizer.Tools.UnitTests/FhirBlobStreamTests.cs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+            }
+
+            writer.Flush();
+
+            stream.Position = 0;
+            await blobClient.UploadAsync(stream);
+            return expectedResult;
+        }
+    }
+}
```
Comment (Patrick2011) [inline src/Fhir.Anonymizer.Tools.UnitTests/FhirBlobStreamTests.cs]: We may need to add test cases for the blob stream reader and writer with cross region storage download and upload. So we can stretch network IO on the component with less compute time.

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs [hunk 6]
```diff
@@ -120,43 +136,43 @@ private async Task AnonymizeBlobInJsonFormatAsync(BlobClient inputBlobClient, Bl
             }
         }
 
-        private async Task AnonymizeBlobInNdJsonFormatAsync(BlobClient inputBlobClient, BlobClient outputBlobClient, string blobName)
+        private async Task AnonymizeBlobInNdJsonFormatAsync(BlobClient inputBlobClient, BlockBlobClient outputBlobClient, string blobName)
```
Comment (Patrick2011) [inline src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs]: We also need to cover cross region support on json files, even though we do not support parallel processing on json file yet.

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirAzureConstants.cs (hunk 3 above)
Comment (deepak-ms) [reply src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/FhirAzureConstants.cs]: sounds good

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs (hunk 4 above)
Comment (deepak-ms) [reply src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs]: Should we add a story to track this work?

Comment (skywalker-wu) [reply src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs]: Create a story to backlog for tracking: https://microsofthealth.visualstudio.com/Health/_workitems/edit/73121/

File: src/Fhir.Anonymizer.Tools.UnitTests/FhirBlobStreamTests.cs (hunk 5 above)
Comment (skywalker-wu) [reply src/Fhir.Anonymizer.Tools.UnitTests/FhirBlobStreamTests.cs]: Add a unit test and run for more than n * 100GB data for data transfer task only. Thanks!

File: src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs (hunk 6 above)
Comment (skywalker-wu) [reply src/Fhir.Anonymizer.AzureDataFactoryPipeline/src/DataFactoryCustomActivity.cs]: Sure, we ran 10 rounds of synthea files json format of cross region scenario. Thanks for the suggestion.

**Pull request merged**